File: programcreek/js/01/01-09-3sum.js
function threeSum(list) {
    var result = [];

    if (list.length < 3) return result;

    list.sort();

    for (var i = 0 ; i < list.length - 2; i ++ ) {
        if (i == 0 || list[i] != list[i - 1]) {

            var negate = - list[i];

            var start = i + 1;
            var end = list.length - 1;

            while (start < end) {

                if ( list[start] + list[end] == negate ) {
                    result.push([list[i], list[start], list[end]]);

                    start ++;
                    end --;

                    while (start < end && list[start] == list[start - 1]) start ++;
                    while (start < end && list[end] == list[end + 1]) end --;
                }
                else if (list[start] + list[end] > negate) {
                    end --;
                }
                else {
                    start ++;
                }

            }


        }
    }

    return result;
}


console.log(threeSum([-1, 0, 1, 2, -1, -4, 1]));

File: programcreek/python/01/01-09-3sum.py

def threeSum(nums) :
    result = []
    if len(nums) < 3:
        return result

    nums.sort()

    for i in range(0, len(nums) - 2) :

        if i == 0 or nums[i] > nums[i - 1]:

            negate = - nums[i]

            start = i
            end = len(nums) - 1

            while start < end:

                if nums[start] + nums[end] == negate:
                    result.append([nums[i], nums[start], nums[end]])

                    start += 1
                    end -= 1

                    while nums[start] == nums[start - 1]:
                        start += 1

                    while nums[end] == nums[end + 1]:
                        end -= 1



                elif nums[start] + nums[end] > negate:
                    end -= 1
                else:
                    start += 1

    return result

print threeSum([-1, 0, 1, 2, -1, -4])